### ISSUE TITLE
Objectloop optimization

### ISSUE THREAD

**Issue by fredrikr** [DavidKinder/Inform6#311]
Note: I only examined this for Z-code.

Objectloop generally loops over all objects in the game. Any checks that are made in the objectloop are made to all objects.

However, if the source code follows this pattern, it changes this to start with child(y) and move to the next with sibling(y);

```
objectloop(x in y) ...
```
Unfortunately, this faster way of iterating over the contents of an object is not used when the code looks like this:

```
objectloop(x in y && x has scenery) ...
```
(or, as far as I can tell, any other condition is added)

Is this something that could and should be changed?

Iterating using sibling is faster and visits the objects in a different order than looping over all objects.

**Comment by erkyrath** [comment]
It's difficult to pattern-match for the general case. E.g. `objectloop(x has scenery && x in y)`. 

Advent.inf has a line

```
objectloop (obj ofclass Treasure && obj in player or location)
```

Also note that using the sibling chain isn't really a safe optimization. E.g. currently this works:

```
objectloop(x in chest && x has treasure)
  move x to player;
```

But if it used the object tree, this would break. 

Cases like this aren't documented as safe, but it's an area where people probably relied what the compiler did rather than what the manual said.

**Comment by heasm66** [comment]
It's unfortunate that the order of the iteration changes, depending on method used. 
```
Attribute treasure;

Object chest "chest";
Object first "first treasure" with has treasure;
Object second "second treasure" has treasure;
Object third "third treasure" with has treasure;

[Main;
	TestObjLoop();
];

[TestObjLoop x;
  ! Mix up order of obj.next
  move first to chest;
  move second to chest;
  move third to chest;
  
  print "In chest, looping over siblings^";
  objectloop(x in chest)  
    print "  ", (name) x, "^";

  print "^In chest, looping over object tree^";
  objectloop(x in chest && x has treasure)
    print "  ", (name) x, "^";

  print "^In chest, looping over siblings again^";
  objectloop(x in chest)
    if (x has treasure) print "  ", (name) x, "^";
];
```
Produces this result
```
In chest, looping over siblings
  third treasure
  second treasure
  first treasure

In chest, looping over object tree
  first treasure
  second treasure
  third treasure

In chest, looping over siblings again
  third treasure
  second treasure
  first treasure
```
There's no easy fix to this, because games can depend on this behaviour.

**Comment by fredrikr** [comment]
I agree that it doesn't make sense to change this behaviour now, as many old games would have weird effects if recompiled.

I *thought* I had seen in DM4 that `if(x in y && ...` would use the sibling-method of iteration, but I must have dreamt this up.
